pyth-sdk/src/price.rs

```diff
@@ -89,6 +90,286 @@ impl Price {
+    /// interpolate/exterpolate to find the value at the query coordinate according to that line
+    /// 
+    /// Args
+    /// x1: i64, the x coordinate of the first point
+    /// y1: Price, the y coordinate of the first point, represented as a Price struct
+    /// x2: i64, the x coordinate of the second point, must be greater than x1
+    /// y2: Price, the y coordinate of the second point, represented as a Price struct
+    /// x_query: the query x coordinate, at which we wish to impute a y value
+    /// 
+    /// Logic
+    /// imputed y value = y2 * ((x3-x1)/(x2-x1)) + y1 * ((x2-x3)/(x2-x1))
+    /// 1. compute A = x3-x1
+    /// 2. compute B = x2-x3
+    /// 3. compute C = x2-x1
+    /// 4. compute D = A/C
+    /// 5. compute E = B/C
+    /// 6. compute F = y2 * D
+    /// 7. compute G = y1 * E
+    /// 8. compute H = F + G
+    /// 
+    /// Bounds due to precision loss
+    /// x = 1/PD_SCALE
+    /// division incurs max loss of x
+    /// Err(D), Err(E) is relatively negligible--by scaling to expo -9 we are imposing
+    /// a grid composed of 1 billion units between x1 and x2 endpoints. Moreover, D, E <= 1.
+    /// Thus, max loss here: Err(D), Err(E) <= x
+    /// Err(y1), Err(y2) often <= x
+    /// Err(F), Err(G) <= (1+x)^2 - 1 (in fractional terms)
+    /// Err(H) <= 2*(1+x)^2 - 2 ~= 2x = 2/PD_SCALE
+    pub fn affine_combination(x1: i64, y1: Price, x2: i64, y2: Price, x_query: i64) -> Result<Price, OracleError> {
+        if x2 <= x1 {
+            return Err(OracleError::InitialEndpointExceedsFinalEndpoint);
+        }
+        
+        // get the deltas for the x coordinates
+        let delta_q1 = x_query - x1;
+        let delta_2q = x2 - x_query;
+        let delta_21 = x2 - x1;
+
+        // convert deltas to Prices
+        let delta_q1_as_price = Price {
+            price: delta_q1,
+            conf: 0,
+            expo: 0,
+            publish_time: 0,
+        };
+        let delta_2q_as_price = Price {
+            price: delta_2q,
+            conf: 0,
+            expo: 0,
+            publish_time: 0,
+        };
+        let delta_21_as_price = Price {
+            price: delta_21,
+            conf: 0,
+            expo: 0,
+            publish_time: 0,
+        };
+
+        // get the relevant fractions of the deltas
+        let mut frac_q1 = delta_q1_as_price.div(&delta_21_as_price).ok_or(OracleError::NoneEncountered)?;
+        let mut frac_2q = delta_2q_as_price.div(&delta_21_as_price).ok_or(OracleError::NoneEncountered)?;
+
+        // scale all the prices to expo -9 for the mul and addition
+        frac_q1 = frac_q1.scale_to_exponent(-9).ok_or(OracleError::NoneEncountered)?;
+        frac_2q = frac_2q.scale_to_exponent(-9).ok_or(OracleError::NoneEncountered)?;
+
+        let y1_scaled = y1.scale_to_exponent(-9).ok_or(OracleError::NoneEncountered)?;
+        let y2_scaled = y2.scale_to_exponent(-9).ok_or(OracleError::NoneEncountered)?;
```

Comment:
another point here is that this will literally be the same logic as `price_basket` if you implement suggestion 2 above. Maybe there's a more composable thing where you have a factory method to generate frac_q1 and frac_q2 and then literally call `price_basket`?